### ISSUE TITLE
Bug in Win32_EntropySource::poll()

### ISSUE THREAD

**Issue by Lexy72** [randombit/botan#1369]
We have updated our Botan Library from 1.10 to 2.3 and noticed a performance degradation. After investigation, we could pinpoint the issue to the Win32_EntropySource::poll(). Here we found that the this specific piece of code in Win32_EntropySource::poll() takes a long time:
   ```
   if(::Heap32ListFirst(snapshot, &heap_list))
         {
         do
            {
            rng.add_entropy_T(heap_list);

            HEAPENTRY32 heap_entry;
            heap_entry.dwSize = sizeof(HEAPENTRY32);
            if(::Heap32First(&heap_entry,
                             heap_list.th32ProcessID,
                             heap_list.th32HeapID))
               {
               do
                  {
                  rng.add_entropy_T(heap_entry);
                  bits += 4;
                  } while(::Heap32Next(&heap_entry));
               }

		 if(bits >= 256)
		    break;

            } while(::Heap32ListNext(snapshot, &heap_list));
         }
      }

```
When looking at the old implementation it seems that in the inner while loop a stop statement is  missing, so I changed it to:
```
              do
              {
                    rng.add_entropy_T(heap_entry);
                    bits += 4;

		      if (bits >= 256) //goal achieved
		        break;
              } while(::Heap32Next(&heap_entry));
```

This solves the performance degradation, and I think the change is ok, but I’m not 100% sure.

**Comment by randombit** [comment]
This change seems ok, especially as we're only relying this source for last ditch protection if the system PRNG fails catastrophically. I will fix this on master shortly. Thanks for the report.

**Comment by randombit** [comment]
Fix will be available in 2.4.0 which is being released ~ January 8th.
